torch/_refs/__init__.py

```diff
@@ -3473,6 +3474,26 @@ def diag(
         return torch.diagonal_copy(self, offset)
 
 
+@register_decomposition(torch.ops.aten.diagonal_scatter)
+@out_wrapper()
+def diagonal_scatter(
+    input: TensorLikeType,
+    src: TensorLikeType,
+    offset: int = 0,
+    dim1: int = 0,
+    dim2: int = 1,
+) -> TensorLikeType:
+    out = input.clone()
+    diag = out.diagonal(offset, dim1, dim2)
+    check(
+        diag.shape == src.shape,
+        lambda: "expected src to have a size equal to the diagonal of the input."
+        f"Got {src.shape} for a diagonal of shape {diag.shape}",
+    )
+    prims.copy_to(diag, src)
```

Comment:
Do `input` (from which `diag` is derived) and `src` have to have the same dtype? I wonder if the `copy_to` prim needs to have its documentation clarified that it expects the tensors to have the same shape and same dtype. The current meta checks for it are odd, and one is commented out (noting it should be moved to a reference). Since the prim was written we decided to move error checks to refs instead of prims wherever possibly, so it's not surprising there's some cleanup to do there.

Reply:
In core it does `diag.copy_(src)` directly. Do we have anything like this in PrimTorch or how should I write it?

Reply:
What about `refs.copy_to`? 

https://github.com/pytorch/pytorch/blob/9ad1659b17ff12109b7bf4e8669d1e07ed4a84e7/torch/_refs/__init__.py#L1724